vignettes/s3-vector.Rmd

```diff
@@ -940,27 +931,223 @@ vec_arith.vctrs_meter.MISSING <- function(op, x, y, ...) {
+
+Remember to run `devtools::document()` to re-generate the `NAMESPACE`, and then take a quick look to familiarise yourself with what is generated. You should see this function appearing in two places (if you've never looked at `NAMESPACE` before note that roxygen2 puts it in alphabetical order):
+
+```{r, eval = FALSE}
+S3method(vec_ptype2,pizza_percent)
+...
+export(vec_ptype2.pizza_percent)
 ```
 
-You also need to register the individual double-dispatch methods. Again, this is harder than it should be because roxygen's heuristics aren't quite right. That means you need to describe the `@method` explicitly:
+Now we finish the work around `vec_ptype2()` by providing all needed methods. Each method has both a `@method` and an `@export` directive. These are purely methods so they don't need to be documented in an `.Rd` file.
 
 ```{r}
-#' @method vec_cast.binned double
+#' @method vec_ptype2.vctrs_percent default
+#' @export                                 
+vec_ptype2.vctrs_percent.default <- function(x, y,
+                                             ...,
+                                             x_arg = "x", y_arg = "y") {
+  vec_default_ptype2(x, y, x_arg = x_arg, y_arg = y_arg)
+}
+
+#' @method vec_ptype2.vctrs_percent vctrs_percent
+#' @export                                       
+vec_ptype2.vctrs_percent.vctrs_percent <- function(x, y, ...) new_percent()
+
+#' @method vec_ptype2.vctrs_percent double       
+#' @export                                       
+vec_ptype2.vctrs_percent.double <- function(x, y, ...) double()
+
+#' @method vec_ptype2.double vctrs_percent
 #' @export
-vec_cast.binned.double <- function(x, y, ...) {
+vec_ptype2.double.vctrs_percent <- function(x, y, ...) double()
+```
+
+The exact same treatment applies to `vec_cast()`, which yields:
+
+```{r eval = FALSE}
+#' @method vec_cast pizza_percent
+#' @export vec_cast.pizza_percent
+#' @export                         
+#' @rdname pizza-vctrs             
+vec_cast.pizza_percent <- function(x, to, ...) {
+  UseMethod("vec_cast.pizza_percent")
 }
+#' @method vec_cast.pizza_percent default
+#' @export 
+vec_cast.pizza_percent.default <- function(x, to, ...) vec_default_cast(x, to)
+#' @method vec_cast.pizza_percent pizza_percent
+#' @export 
+vec_cast.pizza_percent.pizza_percent <- function(x, to, ...) x
+#' @method vec_cast.pizza_percent double
+#' @export
+vec_cast.pizza_percent.double <- function(x, to, ...) percent(x)
+#' @method vec_cast.double pizza_percent
+#' @export
+vec_cast.double.pizza_percent <- function(x, to, ...) vec_data(x)
 ```
 
-Hopefully future versions of roxygen will make these exports less painful.
+For simple casting and coercion functions, this is the recommended formatting. If everything fits on one line, there's no need to use `{}` or add blank lines between method definitions. This compact formatting reinforces the common purpose of these methods and their documentation.
```

Comment:
I recommend the multi-line formatting. Otherwise the bodies don't align since the methods have names of varying width. This makes it harder to scan definitions and easier to get things wrong. Since it's hard to get good test coverage of behaviour for methods, eye-scanning and eye-checking are important.

The single-line formatting also makes it harder to change the code once a method becomes more complicated. And then you get inconsistent formatting in the file since once method has a multi-line `{ }` body and the others haven't.

Reply:
This is a @Hadley opinion.

Reply:
I weakly agree with @lionel-. Pragmatically, if one uses styler, for example, it's not so easy to have these one-off exceptions for when we prefer a different style.

Reply:
I don't care enough about this to fight y'all on it. Feel free to change.

Reply:
@lionel- I leave this in your hands. Obviously I'm OK with the advice, because I had a chance to push back. But I see your point.

Feel free to format in your preferred way. And let's remove any specific formatting advice.

Reply:
ok let's keep the formatting as is and just remove the recommendation, since we don't have a consensus on preferred formatting.

Reply:
I removed the recommendation, so I think we've resolved all the discussion points here. Agree that, at some point, splitting this vignette into topics will make sense. But it's still good have a start on this.